ClosedXML/Excel/Style/XLFill.cs

```diff
@@ -5,142 +5,163 @@ namespace ClosedXML.Excel
 {
     internal class XLFill : IXLFill
     {
-        #region IXLFill Members
+        #region static members
 
-        public bool Equals(IXLFill other)
+        internal static XLFillKey GenerateKey(IXLFill defaultFill)
         {
-            return
-                _patternType == other.PatternType
-                && _backgroundColor.Equals(other.BackgroundColor)
-                && _patternColor.Equals(other.PatternColor)
-                ;
+            XLFillKey key;
+            if (defaultFill == null)
+            {
+                key = XLFillValue.Default.Key;
+            }
+            else if (defaultFill is XLFill)
+            {
+                key = (defaultFill as XLFill).Key;
+            }
+            else
+            {
+                key = new XLFillKey
+                {
+                    PatternType = defaultFill.PatternType,
+                    BackgroundColor = defaultFill.BackgroundColor.Key,
+                    PatternColor = defaultFill.PatternColor.Key
+                };
+            }
+            return key;
         }
 
-        #endregion IXLFill Members
+        #endregion static members
+
+        #region Properties
+
+        private readonly XLStyle _style;
+
+        private XLFillValue _value;
 
-        private void SetStyleChanged()
+        internal XLFillKey Key
         {
-            if (_container != null) _container.StyleChanged = true;
+            get { return _value.Key; }
+            private set { _value = XLFillValue.FromKey(value); }
         }
 
-        public override bool Equals(object obj)
+        #endregion Properties
+
+        #region Constructors
+
+        /// <summary>
+        /// Create an instance of XLFill initializing it with the specified value.
+        /// </summary>
+        /// <param name="style">Style to attach the new instance to.</param>
+        /// <param name="value">Style value to use.</param>
+        public XLFill(XLStyle style, XLFillValue value)
         {
-            return Equals((XLFill)obj);
+            _style = style ?? XLStyle.CreateEmptyStyle();
+            _value = value;
         }
 
-        public override int GetHashCode()
+        public XLFill(XLStyle style, XLFillKey key) : this(style, XLFillValue.FromKey(key))
         {
-            return BackgroundColor.GetHashCode()
-                   ^ (Int32)PatternType
-                   ^ PatternColor.GetHashCode();
         }
 
-        #region Properties
+        public XLFill(XLStyle style = null, IXLFill d = null) : this(style, GenerateKey(d))
+        {
+        }
+
+        #endregion Constructors
+
+        private void Modify(Func<XLFillKey, XLFillKey> modification)
```

Comment:
I notice that the `Modify` methods would recursively call the `modification` method, and it leads to unnecessary, multiple calls of `modification`. Example: `ws.Range("A1:A10").Style.Fill.BackgroundColor = XLColor.Red;` would make `XLFill.cs:125` ( `{ k.PatternType = value; return k; }` ) be called multiple times.

Reply:
```Modify``` is called twice, and that's so due to this code:
```
                // 4 ways of determining an "empty" color
                if (new XLFillPatternValues[] { XLFillPatternValues.None, XLFillPatternValues.Solid }.Contains(PatternType)
                    && (BackgroundColor == null
                    || !BackgroundColor.HasValue
                    || BackgroundColor == XLColor.NoColor
                    || BackgroundColor.ColorType == XLColorType.Indexed && BackgroundColor.Indexed == 64))
                {
                    PatternType = value.HasValue ? XLFillPatternValues.Solid : XLFillPatternValues.None;
                }

                Modify(k => { k.BackgroundColor = value.Key; return k; });
```

When we set a solid background, we change the pattern type too. Not a big overhead.
Or do I understand you wrong?

Reply:
I see. That's actually a side effect of preserving an existing API. 
When we modify a pattern color of a ```XLFill``` from inside the code we can't say what exactly is going on:
**option 1**
```
var f = something.Fill;
f.PatternColor = color;
```
**option 2**
```
var s = something.Style;
s.Fill.PatternColor = color;
```
**option 3**
```
var r = something.AsRange();
r.Style.Fill.PatternColor = color;
```

All these cases must work the same way and produce the same result. So, we do the following
1. Modify the ```Key``` of current ```XLFill``` instance - in case it has been captured by some variable the user would expect it will change.
2. Modify the ```PatternColor``` of the fill of the corresponding style. ```XLStyle.Fill``` creates a new instance each time it is accessed, so it is never the same instance as on step 1. But in case Style is captured by some variable the user again will expect it to be modified,
3. Then, modify ```PatternColor``` of all the fills in the ```_container``` if it exists, and all nested containers too. In your example, they all have the same style (range and all cells in it), so this step is executed only once. But if you apply some other styles to some of the cells there will be even more repetitions.

I admit this may not seem very optimal, but I really tried to make it as simple as possible without modifying existing tests ).